### ISSUE TITLE
Improving Standardization of File Types for Better Pipeline Modularity 

### ISSUE THREAD

**Issue by gottacatchenall** [GEO-BON/bon-in-a-box-pipelines#207]
Didn't quite make it to this in the meeting yesterday, but something I think will be very useful for making modular development of future scripts easy is clear standardization for types of data that are widely used across BiaB scripts.

As a motivating example, consider building an SDM pipeline where we have two different options for both pseudoabsences generation and for crossvalidation splits, so a given pipeline maker decides to follow one pair of the dashed line options below.

```mermaid
flowchart TD
    E[pull presences from GBIF] -.-> F[pseudoabsence method 1]
    E -.-> G[pseudoabsence method 2]
    F --> D
    G --> D
    D([presences_and_absences.csv]) -.-> A
    D -.-> B
    A[crossvalidation method 1]
    B[crossvalidation method 2]
    A --> C([crossvalidation_split.csv])
    B --> C
    C --> H[Fit model]
  ```

For each of these scripts to operate independently of the rest, the output file from each pseudoabsence method (`presences_and_absences.csv`) should have the same column structure, that way all crossvalidation scripts know what the format of the input file is (beyond being just a CSV).  Similarly, the format of `crossvalidation_split.csv` should be standardized so the _Fit Model_ script for any given SDM works with any CV method. 

I think we should decide on a set of standards for commonly used forms of data sooner rather than later. 

# Proposed example standards

Below are a few examples of what I mean by standardized formats. I'm sure there are many types of common data i/os that I'm forgetting

## Occurrence (`foo.occurrence`)

Any set of presences and true/pseudoabsences associated each associated with coordinates. 

| Occurrence | latitude | longitude |
| :---------------- | :------: | ----: |
| 0 |  38.92  | -143.2 |
| 1 |  36.43   | -160.4 |
| 1 |  49.83   | -153.2 |
| 0 |  52.42 | -154.3 |

## Crossvalidation (`foo.occurrence.cv`)

A crossvalidation split applied to occurrence data, with the sets labeled as one of `"test"`, `"train"`, or "`validation`"

| Occurrence | set | latitude | longitude |
| :---------------- | :------: | :------: | ----: |
| 0 | test |  38.92  | -143.2 |
| 1 | train |  36.43   | -160.4 |
| 1 | validation |  49.83   | -153.2 |
| 0 | train |  52.42 | -154.3 |

## Features and Labels (`foo.fnl`)

A set of responses (`y#`) and predictors (`x#`) to be used by a model

| y1 | y2 | x1 | x2 | x3 | x4 |
| :---------------- | :------: | :------: | :------: | :------: | ----: |
| 1.23 | 3.4 | 12.54 | -34.4 | 50.1 | 30.1 |
| -0.31 | 2.1 | 10.23 | -24.3 | 17.3 | 28.4 |

**Comment by jmlord** [comment]
@JoryGriffith @glaroc Agree with Michael we should aim to standardize more. My first question would be, can we rely on existing and widely accepted standards almost exclusively? Second question: should we create a small "standardization task-force" that would check those types of question, along with script redundancy, classification, etc..

The opposite of this: at some point I was thinking the script author could be able to specify mandatory columns in a csv file. But that adds a burden that can be redundant, and then the author would also need to choose the value format for the columns. It brings a lot of fake incompatibilities if columns are named differently, etc. So obviously I did not go that way.

**Comment by JoryGriffith** [comment]
I also agree that we should standardize more! We talked about this with Yanina last week about having a "pipeline standards" document similar to the software standards one that they have on the B-cubed website, and having standardized file types would be an important part of that. This would help with the peer review process as well. It seems like this should be more of a living document though because new things will probably come up as new pipelines are added.

**Comment by tpoisot** [comment]
> Second question: should we create a small "standardization task-force" that would check those types of question, along with script redundancy, classification, etc..

I think that's a good idea, and I'm volunteering if we go forward with this.

**Comment by glaroc** [comment]
I agree with the concept of standardization of specific files like occurrences. For geospatial files, we would ideally choose a file format that allows different CRS and for which the CRS can be obtained from the file, such as GPKG. The other advantage is that a GPKG file can be read through range-requests meaning that certain rows can be extracted without putting the entire file in memory. However, since GPKG files can contain multiple layers of different types, we'd still have to make sure the layer name and column names are standard.
